c8d/prune: Fix images being deleted when they're still used with a different reference

vvoland opened the pull request:
- Fixes: https://github.com/moby/moby/issues/45732


**- What I did**
Fixed `image prune -a` deleting dangling images used by containers.

**- How I did it**
-  Exclude dangling tag of the images used by containers (this makes sure that if the original image was deleted we don't delete the dangling image)
-  Exclude images that are used by containers which were started by specifying an image ID instead of a specific name.

**- How to verify it**
```bash
$ make TEST_INTEGRATION_USE_SNAPSHOTTER=1 \
       DOCKER_GRAPHDRIVER=overlayfs \
       TEST_FILTER=TestPruneDontDeleteUsedDangling \
       test-integration
...
INFO: Testing against a local daemon
=== RUN   TestPruneDontDeleteUsedDangling
--- PASS: TestPruneDontDeleteUsedDangling (0.07s)
PASS

DONE 1 tests in 1.079s
```

**- Description for the changelog**
```release-notes
containerd integration: `docker image prune -a` no longer removes images still in use by running containers
```

**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
testutil/environment: Add GetTestDanglingImageId

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

laurazard reviewed (approved):
LGTM

thaJeztah commented:
Looks like there's a failure;

```
=== FAIL: github.com/docker/docker/integration/image TestPruneDontDeleteUsedDangling (0.03s)
    prune_test.go:23: assertion failed: error is not nil: Error response from daemon: No such image: sha256:0df1207206e5288f4a989a2f13d1f5b3c4e70467702c1d5d21dfc9f002b7bd43
```

thaJeztah commented:
I'm wondering though; if 2 images are tags for the same digest, wouldn't we have to use the reverse? Untag the image, and let containerd prune / garbage collect unused content?

(If we use "force", them probably we must change the tag name to a dangling image name instead)

thaJeztah commented:
Discussing with @neersighted and we were considering if we should take a different approach in general;

- whenever we create an image, always create two images:
    - 1 image with the name:tag (if any)
    - 1 "dangling" image
- the second one is effectively replacing the graph driver "content addressable" store (access image by digest)


Now when we remove an image; we either remove the image's `name:tag` from containerd's store, or the dangling image name (if there's no more references to it).

So instead of creating dangling image references "on the fly", we can create them up-front, and use them for reference counting on our side

vvoland committed:
c8d/softDelete: Extract ensureDanglingImage

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/prune: Exclude dangling tag of the images used by containers

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/prune: Handle containers started from image id

If an image is only by id instead of its name, don't prune it
completely. but only untag it and create a dangling image for it.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
Ah, the failure is on Windows only, because the `hack/make/.build-empty-images` doesn't run on Windows, so there's no special dangling image.

> So instead of creating dangling image references "on the fly", we can create them up-front, and use them for reference counting on our side.

I thought about it before, but it seemed to be a bit more tricky because the image creation is not always on our side (for example building image with Buildkit) whereas it's more common for Moby to perform image deletion.

On the other hand, that would indeed give us more control over the GC and would make more sense in cases where an explicit image digest is used instead of a name (so for example `docker run 3abd238` wouldn't pick **any** image that points to the `sha256:3abd238...`).

We could give it a try and see if that doesn't complicate things too much. It's a bit orthogonal to this PR though, so possibly we should just fix this issue here and then continue in a follow-up/separate ticket.

> I'm wondering though; if 2 images are tags for the same digest, wouldn't we have to use the reverse? Untag the image, and let containerd prune / garbage collect unused content?

Yes, this will cause the unused tag which also points to the same target as the used tag.
Looks like this can be fixed easily by just always excluding the dangling image of the container image.

Also, there's another unsolved issue -  images referenced by their id (`docker run 2376a0c12759aa` or `docker run sha256:...`) will be pruned. 

Fixed it now.

thaJeztah commented on daemon/containerd/image_prune.go:
```diff
@@ -107,12 +125,25 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
+			}
+		}
+		imagesWithDigest[dgst] = totalCount - 1
 	}
```
Somewhat wondering (but need to revisit the logic again) if we  could capture all of these in a single struct; something like;

```go
func prune() {
	// references to this digest, and whether they should be pruned.
	imageRefs := map[digest.Digest]map[string]bool{}

	imageRefs["sha256:deadbeef"] = map[string]bool{}

	// hello-world:latest can be pruned
	imageRefs["sha256:deadbeef"]["docker.io/library/hello-world:latest"] = true

	// or not ... something is still referencing it
	imageRefs["sha256:deadbeef"]["docker.io/library/hello-world:latest"] = false

	// container using the image, but by digest
	imageRefs["sha256:deadbeef"][danglingImageName("sha256:deadbeef")] = false

	for dgst, refs := range imageRefs {
		for ref, prune := range refs {
			if prune {
				// DELETE IMAGE
				// remove this reference from the digest
				delete(imageRefs[dgst], ref)
			}
		}
		if len(imageRefs[dgst]) == 0 {
			// no more references to this digest: DELETE CONTENT
		}
	}
}
```

vvoland replied:
Changed to decrement first.

thaJeztah replied:
I'll "unresolved" this one for future reference (if we're gonna look at possibly restructuring things)

neersighted commented on daemon/containerd/image_prune.go:
```diff
@@ -107,11 +125,27 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
+	for _, img := range imagesToPrune {
+		dgst := img.Target.Digest
+
+		totalCount := imagesWithDigest[dgst]
+		if totalCount == 1 {
+			if _, isUsed := usedDigests[dgst]; isUsed {
+				if err := i.ensureDanglingImage(ctx, img); err != nil {
+					return &report, errors.Wrapf(err, "failed to create ensure dangling image for %s", img.Name)
+				}
+			}
+		}
+		imagesWithDigest[dgst] = totalCount - 1
```
```suggestion
		if imagesWithDigest[dgst] == 1 {
			if _, isUsed := usedDigests[dgst]; isUsed {
				if err := i.ensureDanglingImage(ctx, img); err != nil {
					return &report, errors.Wrapf(err, "failed to create ensure dangling image for %s", img.Name)
				}
			}
		}
		imagesWithDigest[dgst] -= 1
```

neersighted reviewed (changes requested):
Some initial nits, though I'm not a fan of how imperative this code is, and I think I might send a PR to your branch with a restructured implementation that keeps each category of prune exclusion more discrete.

vvoland committed:
c8d/softDelete: Deep copy Labels

So we don't override the original Labels in the passed image object.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on integration/image/inspect_test.go:
```diff
@@ -19,10 +19,7 @@ func TestImageInspectEmptyTagsAndDigests(t *testing.T) {
-	if testEnv.UsingSnapshotter() {
-		danglingId = environment.DanglingImageIdSnapshotter
-	}
+	danglingId := environment.GetTestDanglingImageId(testEnv)
```
Changed, it's a small price for making your OCD side happy 😄

vvoland commented on integration/image/prune_test.go:
```diff
@@ -0,0 +1,38 @@
+		container.WithImage(danglingId),
+		container.WithCmd("sleep", "60"))
+
+	t.Cleanup(func() {
+		client.ContainerRemove(ctx, cid, types.ContainerRemoveOptions{Force: true})
+	})
```
Removed 👍🏻

vvoland commented on integration/image/prune_test.go:
```diff
@@ -0,0 +1,38 @@
+
+// Regression test for: https://github.com/moby/moby/issues/45732
+func TestPruneDontDeleteUsedDangling(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows")
```
Done.

vvoland commented on daemon/containerd/image_prune.go:
```diff
@@ -70,35 +71,52 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
-			imagesToPrune[img.Name] = img
-		}
-	}
+		count := imagesWithDigest[img.Target.Digest]
+		imagesWithDigest[img.Target.Digest] = count + 1
```
Done.

vvoland commented on daemon/containerd/image_prune.go:
```diff
@@ -70,35 +71,52 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
 		return nil, err
 	}
 
+	// How many images there are with each digest.
+	imagesWithDigest := map[digest.Digest]int{}
```
Done.

vvoland commented on daemon/containerd/image_prune.go:
```diff
@@ -70,35 +71,52 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
 	// Exclude images used by existing containers
-	for _, ctr := range containers {
+	for _, ctr := range i.containers.List() {
 		// Config.Image is the image reference passed by user.
```
Done.

vvoland commented on daemon/containerd/image_prune.go:
```diff
@@ -70,35 +71,52 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
+		//           `docker run 82d1e9d` will have Config.Image="82d1e9d"
+		// but both will have ImageID="sha256:82d1e9d7ed48a7523bdebc18cf6290bdb97b82302a8a9c27d4fe885949ea94d1"
+		imageDgst := ctr.ImageID.Digest()
+		truncatedId := strings.HasPrefix(imageDgst.String(), "sha256:"+strings.TrimPrefix(ctr.Config.Image, "sha256:"))
+
+		// If the original image was deleted, make sure we don't delete the dangling image
+		delete(imagesToPrune, danglingImageName(ctr.ImageID.Digest()))
+
+		if truncatedId {
+			// If user didn't specify an explicit image, mark the digest as used.
+			usedDigests[imageDgst] = struct{}{}
+			continue
+		}
```
Changed to `normalizedImageID`.

neersighted commented on daemon/containerd/image_prune.go:
```diff
@@ -70,35 +71,50 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
+		imageDgst := ctr.ImageID.Digest()
+
+		// If user didn't specify an explicit image, mark the digest as used.
+		normalizedImageID := strings.HasPrefix(imageDgst.String(), "sha256:"+strings.TrimPrefix(ctr.Config.Image, "sha256:"))
+		if normalizedImageID {
```
This still doesn't quite make sense -- the intention was to split the normalization from the check to make the intention clear:
```suggestion
		normalizedImageID := "sha256:"+strings.TrimPrefix(ctr.Config.Image, "sha256:")
		if strings.HasPrefix(imageDgst.String(), normalizedImageID) {
```

vvoland replied:
Oh, sorry! I read the comment diff too quickly and assumed the wrong thing 😅 
Fixed now!

neersighted reviewed (approved):
One nit, but overall LGTM. I think that we should restructure this logic in the future, but I think that we can take this as an improvement for now, and think about lifecycle management more holistically in the future (during which we can revisit this logic, and perhaps even generalize it).

thaJeztah commented on integration/image/inspect_test.go:
```diff
@@ -19,9 +19,9 @@ func TestImageInspectEmptyTagsAndDigests(t *testing.T) {
 	ctx := context.Background()
 
-	danglingId := environment.GetTestDanglingImageId(testEnv)
+	danglingID := environment.GetTestDanglingImageId(testEnv)
```
Was this meant to go in the first commit? 🙈 

```diff
	danglingID := environment.GetTestDanglingImageID(testEnv)
```

vvoland replied:
🙈  Fixing it up!

vvoland replied:
Done, I mistakenly did a `--fixup` of both tests to the same commit 😅

thaJeztah reviewed (approved):
LGTM (but let me know if you want to address that nit; not a blocker)

thaJeztah reviewed (approved):
LGTM

neersighted commented:
Backported at https://github.com/moby/moby/pull/45857

thaJeztah commented:
@neersighted commits still align? I see there may be a potential race condition between you cherry-picking and Pawel force-pushing 😓 🙈

neersighted commented:
That's one of the things to check when reviewing a backport PR 😉 

But yes, I made sure to re-cherry (with rerere) once Pawel pushed again; all SHAs are aligned to the latest version of this PR.

thaJeztah commented:
Thanks!

thaJeztah commented:
Timeout in one of the BuildKit tests (panic: test timed out after 30m0s); kicked it again
